<a href="https://colab.research.google.com/github/sv2639/proj3/blob/main/proj3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Problem 1: Explore Different Modes of Operation Through Manual Encryption and Decryption

## Problem 1: Encryption Modes

### Part 2a - Mode: ECB (Electronic Codebook)
Convert the plaintext **FOO** into binary using the provided table:
- **F** = 0101  
- **O** = 1110  
- **O** = 1110  

Apply the block cipher **Ek** to each block:
- **Ek(0101) = 1001**  
- **Ek(1110) = 1110**  
- **Ek(1110) = 1110**  

Convert the ciphertext binary back into letters using the table:
- **1001 = J**  
- **1110 = O**  

Final output:  
**FOO → JOO**  

---

### Part 2b - Mode: CBC (Cipher Block Chaining) with IV = 1010
#### Step 1: Encrypt each block  
1. **Block 1**:  
   - XOR with Initialization Vector:  
     - **0101 ⊕ 1010 = 1111**  
   - Apply the cipher **Ek** on the block:  
     - **Ek(1111) = 1111**  

2. **Block 2**:  
   - XOR with previous ciphertext:  
     - **1110 ⊕ 1111 = 0001**  
   - Apply **Ek**:  
     - **Ek(0001) = 0001**  

3. **Block 3**:  
   - XOR with previous ciphertext:  
     - **1110 ⊕ 0001 = 1111**  
   - Apply **Ek**:  
     - **Ek(1111) = 1111**  

Convert **1111 0001 1111** back into letters using the table:  
**FOO → PBP**  

---

### Part 2c - Mode: CTR (Counter) with ctr = 1010
1. Encrypt the counter:  
   - **CTR = 1010**  
   - Apply **Ek**:  
     - **Ek(1010) = 0110**  

2. Encrypt each block:  
   - **Block 1**:  
     - **0101 ⊕ 0110 = 0011**  

   - **Increase counter**:  
     - **CTR = 1011**  
     - **Ek(1011) = 0111**  

   - **Block 2**:  
     - **1110 ⊕ 0111 = 1001**  

   - **Increase counter**:  
     - **CTR = 1100**  
     - **Ek(1100) = 1010**  

   - **Block 3**:  
     - **1110 ⊕ 1010 = 0100**  

Convert **0011 1001 0100** back into letters using the table:  
**FOO → DJE**  

---

## Problem 1: Decryption Task

### 1) ECB (Electronic Codebook)
Convert the ciphertext into binary:
- **J** = 1001  
- **O** = 1110  
- **O** = 1110  

Apply the inverse of the encryption function:
- **1001 → 0101 → F**  
- **1110 → 1110 → O**  
- **1110 → 1110 → O**  

Final decrypted output:  
**JOO → FOO**  

---

### 2) CBC (Cipher Block Chaining) with IV = 1010
Convert the ciphertext to binary:
- **P** = 1111  
- **B** = 0001  
- **P** = 1111  

Decrypt using CBC:
1. XOR with Initialization Vector:  
   - **1010 ⊕ 1111 = 0101**  
   - Apply **Ek⁻¹**:  
     - **1001 → F**  

2. **Block 2**:  
   - XOR with previous ciphertext:  
     - **0001 ⊕ 1111 = 1110**  
   - Apply **Ek⁻¹**:  
     - **1110 → O**  

3. **Block 3**:  
   - XOR with previous ciphertext:  
     - **0001 ⊕ 1111 = 1110**  
   - Apply **Ek⁻¹**:  
     - **1110 → O**  

Final decrypted output:  
**PBP → FOO**  

---

### 3) CTR (Counter Mode) with Counter = 1010
Convert ciphertext into binary:
- **DJE** = **0011 1001 0100**  

Counters:  
- **Counter1** = **1010**  
- **Counter2** = **1011**  
- **Counter3** = **1100**  

Decrypt each block:
1. **Encrypt Counter1**:  
   - **Ek(1010) = 0110**  
   - **c1 = D = 0011**  
   - **o1 ⊕ c1 = 0110 ⊕ 0011 = 0101 (F)**  

2. **Encrypt Counter2**:  
   - **Ek(1011) = 0111**  
   - **c2 = J = 1001**  
   - **o2 ⊕ c2 = 0111 ⊕ 1001 = 1110 (O)**  

3. **Encrypt Counter3**:  
   - **Ek(1100) = 1010**  
   - **c3 = E = 0100**  
   - **o3 ⊕ c3 = 1010 ⊕ 0100 = 1110 (O)**  

Final decrypted output:  
**DJE → FOO**  


### Problem 2 Implementing a Meet-in-the-Middle Attack on a Mini Block Cipher

### Project Overview:

This project involves students in the implementation and analysis of a meet-in-the-middle (MITM) attack against a simplified block cipher called mini block cipher based on the idea of the SAES (FYI, it is better to
use the simplified DES). The primary goal is to provide hands-on experience with the MITM attack, showcasing its effectiveness against certain cryptographic algorithms and understanding why modern ciphers like AES are designed to be immune to such attacks.

MITM or Meet-in-the-Middle Attack” is an exhaustive key search attack1. It is a cryptanalytic attack applicable to ciphers based on composition of multiple rounds of substitutions and permutations. It works by finding plaintext-ciphertext pairs that map to the same intermediate value after partial encryption/decryption2.

SAES starts with the key expansion, then works on encryption to get ciphertext and on decryption to recover the plaintext. The key expansion generates three keys. The first key, Key0, is used for the add round key to the plaintext. The second key, Key1, is used to perform Round 1 transformation on state, defined as encrypt_round1(). The third key, Key2, is used to perform Round 2 transformations on state,
defined as encrypt_round2(). For decryption, it is reverse. Key2 is used to perform inverse Round 2 transformations on ciphertext, defined as decrypt_round2(). Key1 is used to perform inverse Round 1
transformations on state, defined as decrypt_round1(). So, the pseudo-code for SAES would be the following. It is assumed the key size is 16 bit.

1. Get (Key0, Key1, Key2) from Key K using the key expansion.

2. Two rounds of Encrypt to get the ciphertext.
  
        I. AddRoundKey Key0
        
        II. encrypt_round1() using Key1
    
            i. Subsititute()
    
            ii. Shift()
    
            iii. Mix()
    
            iv. AddRoundKey()
        
        III. encrypt_round2() using Key2
    
            i. Subsititute()
    
            ii. Shift()
    
            iii. AddRoundKey()

3. Two rounds of decryption to recover the plaintext
  
        I. decrypt_round2() using Key 2
    
            i. AddRoundKey()
    
            ii. Shift()
    
            iii. Subsititute()
  
        II. decrypt_round1() using Key1

            i. AddRoundKey()
            
            ii. Mix()
            
            iii. Shift()
            
            iv. Subsititute()

        III. AddRoundKey Key0

For the project, we need to modify the pseudo code to fit our class project. The pseudo-code for this
mini block cipher based on the SAES is as follows:

1. Get (Key0, Key1, Key2) from Key K using the key expansion.

2. Two rounds of Encrypt to get the ciphertext.

        I. encrypt_round1() using Key1 on plaintext, P, and get intermediate state X

            i. Subsititute()

            ii. Shift()

            iii. Mix()

            iv. AddRoundKey()

        II. encrypt_round2() using Key2 on intermediate state X, and get the ciphertext C.

            i. Subsititute()
  
            ii. Shift()

            iii. AddRoundKey()

3. Two rounds of decryption to recover the plaintext

        I. decrypt_round2() using Key 2 on the ciphertext, C, and get intermediate state, Y

            i. AddRoundKey()

            ii. Shift()
  
            iii. Subsititute()

        II. decrypt_round1() using Key1 on Y to get plaintext P.

            i. AddRoundKey()

            ii. Mix()

            iii. Shift()

            iv. Subsititute()

The meet in the middle attack strategy to mini block cipher:

      A. Calculate X = encrypt_round1(Key1, P)

      B. Calculate X’ = decrypt_round2(Key2, C).

      C. Find out a pair (Key1, Key2) such at X = X’

      D. For one specific (P, C), there will be many matched pairs, we need to use another plaintext and ciphertext pair to eliminate some of the matched pairs. Ideally, we shall get one matched key pair. This pair can be used to get plaintext from any cyphertext. In other words, we cracked the block cipher.

Project Tasks

Task 1: Implementing Mini Block Cipher with key size 16 bit and block size 16 bit:

      a) Students will implement the Mini Block Cipher encryption and decryption functions (1, 2I, 2II, 3I, 3II) using jupyter notebook.

      b) Make at least ten pairs of plaintexts and ciphertexts.

Task 2: Meet-in-the-Middle Attack Implementation:

      a) Students need to implement the meet in the middle attack strategy to mini block cipher (a-d).

      b) Show key pair(s) that works for the pair of plaintext and ciphertext from task1 (b). Ideally, it should have only one key pair works.

Task 3: Analyze the time and memory complexity of the attack compared with the naive exhaustive key search.

      a) What is the key space for the mini block cipher?

      b) Image the mini block cipher is executed twice to generate a cipher text. It is called double mini cipher block. We need a key in 32 bits. The first 16 to the first mini block cipher, the remaining 16 to the second mini block cipher. The meet in the middle attack is to match the state for the first encryption of mini block cipher and the second decryption mini block. How many operations are needed to such attack?

      c) If we do exhaustive key search for the double mini block cipher, how many operations are needed?

      d) What is the tradeoff for the MITM attack (speed, memory, etc.)?





1 https://en.wikipedia.org/wiki/Meet-in-the-middle_attack

2 For example, https://youtu.be/S-EhbhDXUwM (It is a bit long…)

old. Needs to be removed:

In [ ]:
import random

# Define the S-box substitution table
s_box = {
    "0000": "1001", "0001": "0100", "0010": "1010", "0011": "1011",
    "0100": "1101", "0101": "0001", "0110": "1000", "0111": "0101",
    "1000": "0110", "1001": "0010", "1010": "0000", "1011": "0011",
    "1100": "1100", "1101": "1110", "1110": "1111", "1111": "0111"
}

# Define the inverse S-box for correct decryption
inv_s_box = {v: k for k, v in s_box.items()}  # Swap keys and values

# Substitution functions
def substitute(nibble):
    return s_box.get(nibble, "Invalid Input")  # Forward S-box substitution

def inverse_substitute(nibble):
    return inv_s_box.get(nibble, "Invalid Input")  # Correct inverse substitution

# Other helper functions
def shift(block):
    return block[1:] + block[0]  # Circular shift left

def inverse_shift(block):
    return block[-1] + block[:-1]  # Circular shift right (inverse)

def mix(block):
    constant = "1100"  # Example mix operation with XOR
    return ''.join(['1' if b != constant[i] else '0' for i, b in enumerate(block)])

def inverse_mix(block):
    constant = "1100"  # Same constant used in mix
    return ''.join(['1' if b != constant[i] else '0' for i, b in enumerate(block)])  # Undo mix

def add_round_key(block, key):
    return ''.join(['1' if b != k else '0' for b, k in zip(block, key)])  # XOR operation

# Encryption rounds
def encrypt_round1(block, key):
    block = substitute(block)
    block = shift(block)
    block = mix(block)
    block = add_round_key(block, key)
    return block

def encrypt_round2(block, key):
    block = substitute(block)
    block = shift(block)
    block = add_round_key(block, key)
    return block

# **Fixed decryption functions**
def decrypt_round2(block, key):
    block = add_round_key(block, key)
    block = inverse_shift(block)  # Correct inverse shift
    block = inverse_substitute(block)  # Correct inverse S-box substitution
    return block

def decrypt_round1(block, key):
    block = add_round_key(block, key)
    block = inverse_mix(block)  # Correct inverse mix function
    block = inverse_shift(block)  # Correct inverse shift
    block = inverse_substitute(block)  # Correct inverse S-box substitution
    return block

# SAES encryption & decryption
def encrypt(plaintext, keys):
    key0, key1, key2 = keys  # Unpack keys
    ciphertext = ''
    for i in range(0, len(plaintext), 4):
        block = plaintext[i:i+4]
        intermediate = encrypt_round1(block, key1)
        ciphertext += encrypt_round2(intermediate, key2)
    return ciphertext

def decrypt(ciphertext, keys):
    key0, key1, key2 = keys
    plaintext = ''
    for i in range(0, len(ciphertext), 4):
        block = ciphertext[i:i+4]
        intermediate = decrypt_round2(block, key2)
        plaintext += decrypt_round1(intermediate, key1)
    return plaintext

# Character to Binary Mapping
char_to_bin = {
    'A': '0000', 'B': '0001', 'C': '0010', 'D': '0011',
    'E': '0100', 'F': '0101', 'G': '0110', 'H': '0111',
    'I': '1000', 'J': '1001', 'K': '1010', 'L': '1011',
    'M': '1100', 'N': '1101', 'O': '1110', 'P': '1111'
}

bin_to_char = {v: k for k, v in char_to_bin.items()}  # Reverse mapping

def text_to_binary(text):
    return ''.join([char_to_bin[char] for char in text.upper() if char in char_to_bin])

def binary_to_text(binary):
    return ''.join([bin_to_char[binary[i:i+4]] for i in range(0, len(binary), 4)])

# Key Expansion & Random Key Generator
def key_expansion(key):
    return key[:4], key[4:8], key[8:12]  # Split key into three parts

def generate_random_key():
    return ''.join(random.choice('01') for _ in range(12))

# **Testing the Fix**
key = generate_random_key()
key_exp = key_expansion(key)

plaintext = "ABC"  # Example input
plaintext_binary = text_to_binary(plaintext)

print(f"Original Plaintext: {plaintext}")
print(f"Binary Plaintext: {plaintext_binary}")

ciphertext = encrypt(plaintext_binary, key_exp)
print(f"Ciphertext: {ciphertext}")

decrypted_binary = decrypt(ciphertext, key_exp)
decrypted_text = binary_to_text(decrypted_binary)

print(f"Decrypted Binary: {decrypted_binary}")
print(f"Decrypted Plaintext: {decrypted_text}")
print(f"Key: {key}")

assert plaintext == decrypted_text, "Decryption failed!"


Original Plaintext: ABC
Binary Plaintext: 000000010010
Ciphertext: 100010010010
Decrypted Binary: 000000010010
Decrypted Plaintext: ABC
Key: 011101101100


 In the pseudo code for question 2 in project 3, it says the following:
> For the project, we need to modify the pseudo code to fit our class project. The pseudo-code for this
> mini block cipher based on the SAES is as follows:
> 1. Get (Key0, Key1, Key2) from Key K using the key expansion


It says to get key0, key1, and key2

But then steps 2 and 3 say to use key1 and key2, completely skipping key0

The pseudo-code for SAES says to use key0, before either of the encrypt rounds. Meaning key0 actually gets used.

But I don't see where key0 gets used in the mini block cipherSo that's why I'm messaging you.Why does key0 get made if it's not being used?

Secondly. For the substitute portion of the algorithm are we using the substitution box from the slides of this week's lecture?



## Task 1A

old. Needs to be removed:

In [1]:
import random

# Define the S-box substitution table (adjusted for 5-bit input)
s_box = {
    "00000": "11010", "00001": "10011", "00010": "01000", "00011": "10100",
    "00100": "01101", "00101": "11110", "00110": "00111", "00111": "10101",
    "01000": "00010", "01001": "11001", "01010": "01110", "01011": "11100",
    "01100": "10000", "01101": "00110", "01110": "01011", "01111": "00001",
    "10000": "10110", "10001": "10001", "10010": "00011", "10011": "11011",
    "10100": "00101", "10101": "11111", "10110": "01001", "10111": "10010",
    "11000": "01111", "11001": "00000", "11010": "11000", "11011": "10111",
    "11100": "00100", "11101": "01100", "11110": "01010", "11111": "11101"
}

inv_s_box = {v: k for k, v in s_box.items()}  # Reverse mapping

# Substitution functions
def substitute(nibble):
    if nibble not in s_box:
        raise ValueError(f"Invalid S-Box input: {nibble}")
    return s_box[nibble]

def inverse_substitute(nibble):
    if nibble not in inv_s_box:
        raise ValueError(f"Invalid Inverse S-Box input: {nibble}")
    return inv_s_box[nibble]


def shift(block):
    return block[2:] + block[:2]  # Shift by 2

def inverse_shift(block):
    return block[-2:] + block[:-2]  # Reverse shift by 2


def mix(block):
    constant = "11010"  # block size 5 bits
    return ''.join(['1' if b != constant[i] else '0' for i, b in enumerate(block)])

def inverse_mix(block):
    constant = "11010"  # Ensure block size consistency
    return ''.join(['1' if b != constant[i] else '0' for i, b in enumerate(block)])

def add_round_key(block, key):
    return ''.join(['1' if b != k else '0' for b, k in zip(block, key)])  # XOR

# Encryption rounds (updated for 5-bit blocks)
def encrypt_round1(block, key):
    block = substitute(block)
    block = shift(block)
    block = mix(block)
    block = add_round_key(block, key)
    return block

def encrypt_round2(block, key):
    block = substitute(block)
    block = shift(block)
    block = add_round_key(block, key)
    return block

# **Decryption functions (updated for 5-bit blocks)**
def decrypt_round2(block, key):
    block = add_round_key(block, key)
    block = inverse_shift(block)
    block = inverse_substitute(block)
    return block

def decrypt_round1(block, key):
    block = add_round_key(block, key)
    block = inverse_mix(block)
    block = inverse_shift(block)
    block = inverse_substitute(block)
    return block

# **SAES Encryption & Decryption**
def encrypt(plaintext, keys):
    key0, key1, key2 = keys  # Unpack keys
    ciphertext = ''
    for i in range(0, len(plaintext), 5):  # Process full 5-bit blocks
        block = plaintext[i:i+5]
        if len(block) < 5:
            continue  # Ignore incomplete blocks (optional)
        intermediate = encrypt_round1(block, key1)
        ciphertext += encrypt_round2(intermediate, key2)
    return ciphertext

def decrypt(ciphertext, keys):
    key0, key1, key2 = keys
    plaintext = ''
    for i in range(0, len(ciphertext), 5):  # Process full 5-bit blocks
        block = ciphertext[i:i+5]
        if len(block) < 5:
            continue  # Ignore incomplete blocks (optional)
        intermediate = decrypt_round2(block, key2)
        plaintext += decrypt_round1(intermediate, key1)
    return plaintext

# **Character to Binary Mapping (5-bit)**
char_to_bin = {
    'A': '00000', 'B': '00001', 'C': '00010', 'D': '00011',
    'E': '00100', 'F': '00101', 'G': '00110', 'H': '00111',
    'I': '01000', 'J': '01001', 'K': '01010', 'L': '01011',
    'M': '01100', 'N': '01101', 'O': '01110', 'P': '01111',
    'Q': '10000', 'R': '10001', 'S': '10010', 'T': '10011',
    'U': '10100', 'V': '10101', 'W': '10110', 'X': '10111',
    'Y': '11000', 'Z': '11001', ' ': '11010'
}

bin_to_char = {v: k for k, v in char_to_bin.items()}  # Reverse mapping

def text_to_binary(text):
    return ''.join([char_to_bin[char] for char in text.upper() if char in char_to_bin])

def binary_to_text(binary):
    return ''.join([bin_to_char[binary[i:i+5]] for i in range(0, len(binary), 5)])

# **Key Expansion & Random Key Generator (16-bit)**
def key_expansion(key):
    return key[:6], key[6:11], key[11:16]  # 6-bit Key0, 5-bit Key1, 5-bit Key2

def generate_random_key():
    return ''.join(random.choice('01') for _ in range(16))

# **Testing the Fix**
key = generate_random_key()
key_exp = key_expansion(key)

plaintext = "HELLO TEST WORLD"  # Example input
plaintext_binary = text_to_binary(plaintext)

ciphertext = encrypt(plaintext_binary, key_exp)
decrypted_binary = decrypt(ciphertext, key_exp)
decrypted_text = binary_to_text(decrypted_binary)

print(f"Original Plaintext: {plaintext}")
print(f"Binary Plaintext: {plaintext_binary}")
print(f"Ciphertext: {ciphertext}")
print(f"Decrypted Binary: {decrypted_binary}")
print(f"Decrypted Plaintext: {decrypted_text}")
print(f"Key: {key}")

assert plaintext == decrypted_text, "Decryption failed!"


Original Plaintext: HELLO TEST WORLD
Binary Plaintext: 00111001000101101011011101101010011001001001010011110101011001110100010101100011
Ciphertext: 10000011100001000010000000101011100011100100111100010101001000000100010001011110
Decrypted Binary: 00111001000101101011011101101010011001001001010011110101011001110100010101100011
Decrypted Plaintext: HELLO TEST WORLD
Key: 0101101100000100


In [22]:
# KEY EXPANSION


import sys

# AES S-Box (from standard AES specification)
s_box = [
    0x63, 0x7C, 0x77, 0x7B, 0xF2, 0x6B, 0x6F, 0xC5, 0x30, 0x01, 0x67, 0x2B, 0xFE, 0xD7, 0xAB, 0x76,
    0xCA, 0x82, 0xC9, 0x7D, 0xFA, 0x59, 0x47, 0xF0, 0xAD, 0xD4, 0xA2, 0xAF, 0x9C, 0xA4, 0x72, 0xC0,
    0xB7, 0xFD, 0x93, 0x26, 0x36, 0x3F, 0xF7, 0xCC, 0x34, 0xA5, 0xE5, 0xF1, 0x71, 0xD8, 0x31, 0x15,
    0x04, 0xC7, 0x23, 0xC3, 0x18, 0x96, 0x05, 0x9A, 0x07, 0x12, 0x80, 0xE2, 0xEB, 0x27, 0xB2, 0x75,
    0x09, 0x83, 0x2C, 0x1A, 0x1B, 0x6E, 0x5A, 0xA0, 0x52, 0x3B, 0xD6, 0xB3, 0x29, 0xE3, 0x2F, 0x84,
    0x53, 0xD1, 0x00, 0xED, 0x20, 0xFC, 0xB1, 0x5B, 0x6A, 0xCB, 0xBE, 0x39, 0x4A, 0x4C, 0x58, 0xCF,
    0xD0, 0xEF, 0xAA, 0xFB, 0x43, 0x4D, 0x33, 0x85, 0x45, 0xF9, 0x02, 0x7F, 0x50, 0x3C, 0x9F, 0xA8,
    0x51, 0xA3, 0x40, 0x8F, 0x92, 0x9D, 0x38, 0xF5, 0xBC, 0xB6, 0xDA, 0x21, 0x10, 0xFF, 0xF3, 0xD2,
    0xCD, 0x0C, 0x13, 0xEC, 0x5F, 0x97, 0x44, 0x17, 0xC4, 0xA7, 0x7E, 0x3D, 0x64, 0x5D, 0x19, 0x73,
    0x60, 0x81, 0x4F, 0xDC, 0x22, 0x2A, 0x90, 0x88, 0x46, 0xEE, 0xB8, 0x14, 0xDE, 0x5E, 0x0B, 0xDB,
    0xE0, 0x32, 0x3A, 0x0A, 0x49, 0x06, 0x24, 0x5C, 0xC2, 0xD3, 0xAC, 0x62, 0x91, 0x95, 0xE4, 0x79,
    0xE7, 0xC8, 0x37, 0x6D, 0x8D, 0xD5, 0x4E, 0xA9, 0x6C, 0x56, 0xF4, 0xEA, 0x65, 0x7A, 0xAE, 0x08,
    0xBA, 0x78, 0x25, 0x2E, 0x1C, 0xA6, 0xB4, 0xC6, 0xE8, 0xDD, 0x74, 0x1F, 0x4B, 0xBD, 0x8B, 0x8A,
    0x70, 0x3E, 0xB5, 0x66, 0x48, 0x03, 0xF6, 0x0E, 0x61, 0x35, 0x57, 0xB9, 0x86, 0xC1, 0x1D, 0x9E,
    0xE1, 0xF8, 0x98, 0x11, 0x69, 0xD9, 0x8E, 0x94, 0x9B, 0x1E, 0x87, 0xE9, 0xCE, 0x55, 0x28, 0xDF,
    0x8C, 0xA1, 0x89, 0x0D, 0xBF, 0xE6, 0x42, 0x68, 0x41, 0x99, 0x2D, 0x0F, 0xB0, 0x54, 0xBB, 0x16
]

# Generate Inverse S-Box
inv_s_box = [0] * 256
for i in range(256):
    inv_s_box[s_box[i]] = i


Rcon = [
    0x01, 0x02, 0x04, 0x08, 0x10, 0x20, 0x40, 0x80, 0x1B, 0x36 # First 10 Round Constants
]

def sub_byte(byte):
    return s_box[byte]

def inv_sub_byte(byte):
    return inv_s_box[byte]

def rot_word_16bit(word_16bit):
    byte1 = word_16bit & 0xFF  # Least significant byte
    byte2 = (word_16bit >> 8) & 0xFF # Most significant byte
    return (byte2 << 8) | byte1 # Swap bytes

def sub_word_16bit(word_16bit):
    byte1 = word_16bit & 0xFF
    byte2 = (word_16bit >> 8) & 0xFF
    return (sub_byte(byte2) << 8) | sub_byte(byte1)


def inv_sub_word_16bit(word_16bit):
    byte1 = word_16bit & 0xFF
    byte2 = (word_16bit >> 8) & 0xFF
    return (inv_sub_byte(byte2) << 8) | inv_sub_byte(byte1)

def key_expansion(key_16bit):
    """
    Simplified key expansion for a 16-bit key to generate 3 round keys (each 16-bit): Key0, Key1, Key2.
    key_16bit: 16-bit integer representing the AES key.
    Returns: A list of 3 round keys [Key0, Key1, Key2] (each 16-bit integer).
    """
    words = [0] * 3 # We need 3 words W0, W1, W2 which will be Key0, Key1, Key2
    words[0] = key_16bit  # Initialize first word with the key

    for i in range(1, 3): # Generate words W1 and W2
        temp = words[i-1]
        temp_rot = rot_word_16bit(temp)
        temp_sub = sub_word_16bit(temp_rot)
        rc_val = Rcon[i-1] if i <= len(Rcon) else 0 # Get Round Constant for round i (or 0 if out of Rcon range)
        rc_word = rc_val << 8 # Represent RC as 16-bit word (pad with zero byte in lower byte)
        words[i] = words[i-1] ^ temp_sub ^ rc_word

    round_keys = []
    for i in range(3): # Generate 3 round keys Key0 to Key2
        round_keys.append(words[i]) # Keyi = Wi
    return round_keys

if __name__ == '__main__':
    # Example usage with a 16-bit key
    example_key_16bit = 0xABCD # Example 16-bit key

    round_keys = key_expansion(example_key_16bit)

    print("Round Keys (16-bit hex) - Key0, Key1, Key2:")
    print(f"Key0 = 0x{round_keys[0]:04X}")
    print(f"Key1 = 0x{round_keys[1]:04X}")
    print(f"Key2 = 0x{round_keys[2]:04X}")


Round Keys (16-bit hex) - Key0, Key1, Key2:
Key0 = 0xABCD
Key1 = 0xC870
Key2 = 0x2221


In [42]:
#ENCRYPTION


def subsitute_state(state_16bit):
    return sub_word_16bit(state_16bit)

def shift_state(state_16bit):
    return rot_word_16bit(state_16bit)

def mix_state(state_16bit):
    byte1 = state_16bit & 0xFF
    byte2 = (state_16bit >> 8) & 0xFF
    mixed_byte1 = byte2
    mixed_byte2 = byte1 ^ byte2
    return (mixed_byte2 << 8) | mixed_byte1

def add_round_key(state_16bit, round_key_16bit):
    return state_16bit ^ round_key_16bit

def encrypt_round1(plaintext_16bit, key1_16bit):
    state = plaintext_16bit
    state = subsitute_state(state)
    state = shift_state(state)
    state = mix_state(state)
    state = add_round_key(state, key1_16bit)
    return state # Intermediate state X

def encrypt_round2(intermediate_state_16bit, key2_16bit):
    state = intermediate_state_16bit
    state = subsitute_state(state)
    state = shift_state(state)
    state = add_round_key(state, key2_16bit) # No Mix in the last round
    return state # Ciphertext


# Combined Encryption function
def encrypt(plaintext_16bit, key_16bit):
    """
    Encrypts a 16-bit plaintext using the simplified 2-round SAES with a 16-bit key.
    plaintext_16bit: 16-bit integer plaintext.
    key_16bit: 16-bit integer key.
    Returns: 16-bit integer ciphertext.
    """
    round_keys = key_expansion(key_16bit)
    key1, key2 = round_keys[1], round_keys[2] # Use Key1 and Key2 for encryption rounds

    intermediate_state_X = encrypt_round1(plaintext_16bit, key1)
    ciphertext_16bit = encrypt_round2(intermediate_state_X, key2)
    return ciphertext_16bit

In [47]:
#DECRYPTION


def inverse_substitute_state(state_16bit):
    return inv_sub_word_16bit(state_16bit)

def inverse_shift_state(state_16bit):
    return shift_state(state_16bit) # ShiftRows is its own inverse for byte swap

def inverse_mix_state(state_16bit):
    # If mix_state is simplified to: mixed_byte1 = byte1 ^ byte2; mixed_byte2 = byte2;
    # then inverting it: byte2 = mixed_byte2; byte1 = mixed_byte1 ^ byte2 = mixed_byte1 ^ mixed_byte2;
    mixed_byte1 = state_16bit & 0xFF
    mixed_byte2 = (state_16bit >> 8) & 0xFF
    byte2 = mixed_byte1
    byte1 = mixed_byte2 ^ byte2
    return (byte2 << 8) | byte1


def decrypt_round2(ciphertext_16bit, key2_16bit): # Decryption round 2 (first decryption round)
    state = ciphertext_16bit
    state = add_round_key(state, key2_16bit)
    state = inverse_shift_state(state)
    state = inverse_substitute_state(state)
    return state # Intermediate state Y

def decrypt_round1(intermediate_state_16bit, key1_16bit): # Decryption round 1 (second decryption round)
    state = intermediate_state_16bit
    state = add_round_key(state, key1_16bit)
    state = inverse_mix_state(state)
    state = inverse_shift_state(state)
    state = inverse_substitute_state(state)
    return state # Plaintext

# Combined Decryption function
def decrypt(ciphertext_16bit, key_16bit):
    """
    Decrypts a 16-bit ciphertext using the simplified 2-round SAES with a 16-bit key.
    ciphertext_16bit: 16-bit integer ciphertext.
    key_16bit: 16-bit integer key (same key used for encryption).
    Returns: 16-bit integer recovered plaintext.
    """
    round_keys = key_expansion(key_16bit)
    key1, key2 = round_keys[1], round_keys[2] # Use Key1 and Key2 (same keys as in encryption, but in reverse round order in decryption process)

    intermediate_state_Y = decrypt_round2(ciphertext_16bit, key2) # Decryption round 2 uses Key2
    recovered_plaintext_16bit = decrypt_round1(intermediate_state_Y, key1) # Decryption round 1 uses Key1
    return recovered_plaintext_16bit

In [48]:
# Function to generate a random 16-bit key
def generate_random_key_16bit():
    """
    Generates a random 16-bit key.
    Returns: A random 16-bit integer.
    """
    return random.randint(0, 0xFFFF)

In [49]:

# Example usage with a 16-bit key and plaintext
example_key_16bit = generate_random_key_16bit()
plaintext_16bit = 0x1234 # Example 16-bit plaintext

print("Plaintext:", f"0x{plaintext_16bit:04X}")
print("Key:", f"0x{example_key_16bit:04X}")

ciphertext = encrypt(plaintext_16bit, example_key_16bit)
print("\nCiphertext:", f"0x{ciphertext:04X}")

recovered_plaintext = decrypt(ciphertext, example_key_16bit)
print("Recovered Plaintext:", f"0x{recovered_plaintext:04X}")

print("\nVerification:")
if plaintext_16bit == recovered_plaintext:
  print("\n Encryption and Decryption successful! Recovered plaintext matches original plaintext.")
else:
  print("\n Encryption and Decryption failed. Recovered plaintext does NOT match original plaintext.")

Plaintext: 0x1234
Key: 0x5045

Ciphertext: 0x1142
Recovered Plaintext: 0x1234

Verification:

 Encryption and Decryption successful! Recovered plaintext matches original plaintext.


The above code is likely the final version of Task 1a. Unless someone thinks that we should add lower case letters/numbers/special chars. Which could be acheived with slight modification. But I'm not sure it's needed.

Why I think it's the final version:


*   16 bit key
*   Full alphabet
*   Spaces allowed
*   Works without issue





## Task 1B

In [50]:
# Generate a random 16-bit key
example_key_16bit = generate_random_key_16bit()

print("Using Random Key:", f"0x{example_key_16bit:04X}")

plaintext_ciphertext_pairs = {} # Dictionary to store pairs

print("\nGenerating 10 Plaintext-Ciphertext pairs with random key:")
print("---------------------------------------")

for i in range(10):
    # Generate sequential plaintexts for simplicity, you can use random plaintexts as well
    plaintext_16bit = i + 1
    plaintext_hex = f"0x{plaintext_16bit:04X}"

    ciphertext = encrypt(plaintext_16bit, example_key_16bit)
    ciphertext_hex = f"0x{ciphertext:04X}"

    print(f"Pair {i+1}:")
    print(f"  Plaintext:  {plaintext_hex}")
    print(f"  Ciphertext: {ciphertext_hex}")
    print("---------------------------------------")

    # Store the pair in the dictionary
    plaintext_ciphertext_pairs[f"Pair {i+1}"] = {
        "plaintext": plaintext_hex,
        "ciphertext": ciphertext_hex
    }

print("\nVerification of Decryption for the last pair:")
last_plaintext = plaintext_16bit
last_ciphertext = ciphertext
recovered_plaintext = decrypt(last_ciphertext, example_key_16bit)

print(f"Ciphertext to decrypt: 0x{last_ciphertext:04X}")
print(f"Recovered Plaintext:   0x{recovered_plaintext:04X}")

if last_plaintext == recovered_plaintext:
    print("\n Decryption Verified for the last pair: Recovered plaintext matches original.")
else:
    print("\n Decryption Verification Failed for the last pair.")

print("\nPlaintext-Ciphertext Pairs Dictionary:")
print(plaintext_ciphertext_pairs)

Using Random Key: 0xE859

Generating 10 Plaintext-Ciphertext pairs with random key:
---------------------------------------
Pair 1:
  Plaintext:  0x0001
  Ciphertext: 0x0C7C
---------------------------------------
Pair 2:
  Plaintext:  0x0002
  Ciphertext: 0x037C
---------------------------------------
Pair 3:
  Plaintext:  0x0003
  Ciphertext: 0x327C
---------------------------------------
Pair 4:
  Plaintext:  0x0004
  Ciphertext: 0x217C
---------------------------------------
Pair 5:
  Plaintext:  0x0005
  Ciphertext: 0xEA7C
---------------------------------------
Pair 6:
  Plaintext:  0x0006
  Ciphertext: 0xC37C
---------------------------------------
Pair 7:
  Plaintext:  0x0007
  Ciphertext: 0x787C
---------------------------------------
Pair 8:
  Plaintext:  0x0008
  Ciphertext: 0xCD7C
---------------------------------------
Pair 9:
  Plaintext:  0x0009
  Ciphertext: 0xFA7C
---------------------------------------
Pair 10:
  Plaintext:  0x000A
  Ciphertext: 0x087C
---------------

## Task 2A

In [130]:
def mitm_attack(plaintext_16bit1, ciphertext_16bit1, plaintext_16bit2, ciphertext_16bit2):
    """
    Performs a Meet-in-the-Middle attack to find the key pair (Key1, Key2)
    for the simplified 2-round SAES, using TWO plaintext-ciphertext pairs for verification.
    plaintext_16bit1: First known plaintext (16-bit integer).
    ciphertext_16bit1: Corresponding ciphertext for first plaintext (16-bit integer).
    plaintext_16bit2: Second known plaintext (16-bit integer).
    ciphertext_16bit2: Corresponding ciphertext for second plaintext (16-bit integer).
    Returns: A list of potential key pairs [(Key1, Key2), ...] that could have encrypted
             BOTH plaintext pairs to their respective ciphertexts.
    """
    x_values = {} # Dictionary to store X = encrypt_round1(Key1, P1) values (using the first plaintext)
    x_prime_values = {} # Dictionary to store X' = decrypt_round2(Key2, C1) values (using the first ciphertext)

    print("Step A: Pre-calculating X values (using P1)...")
    for key1 in range(0x10000): # Iterate through all possible 16-bit Key1 values
        x = encrypt_round1(plaintext_16bit1, key1)
        if x not in x_values:
            x_values[x] = []
        x_values[x].append(key1)

    print("Step B: Pre-calculating X' values (using C1)...")
    for key2 in range(0x10000): # Iterate through all possible 16-bit Key2 values
        x_prime = decrypt_round2(ciphertext_16bit1, key2)
        if x_prime not in x_prime_values:
            x_prime_values[x_prime] = []
        x_prime_values[x_prime].append(key2)

    matched_key_pairs = []
    print("Step C: Finding matching key pairs...")
    for x_val, key1_list in x_values.items():
        if x_val in x_prime_values:
            for key1 in key1_list:
                for key2 in x_prime_values[x_val]:
                    matched_key_pairs.append((key1, key2))

    print(f"Step C: Found {len(matched_key_pairs)} potential key pairs.")
    return matched_key_pairs



def verify_key_pair(key_pair, plaintext_16bit1, ciphertext_16bit1, plaintext_16bit2, ciphertext_16bit2):
    """
    Verifies if a key pair correctly encrypts BOTH plaintext pairs to their respective ciphertexts.
    key_pair: A tuple (Key1, Key2).
    plaintext_16bit1: First plaintext (16-bit integer).
    ciphertext_16bit1: First ciphertext (16-bit integer).
    plaintext_16bit2: Second plaintext (16-bit integer).
    ciphertext_16bit2: Second ciphertext (16-bit integer).
    Returns: True if the key pair is correct for both pairs, False otherwise.
    """
    key1, key2 = key_pair
    calculated_intermediate_X1 = encrypt_round1(plaintext_16bit1, key1)
    calculated_ciphertext1 = encrypt_round2(calculated_intermediate_X1, key2)
    calculated_intermediate_X2 = encrypt_round1(plaintext_16bit2, key1)
    calculated_ciphertext2 = encrypt_round2(calculated_intermediate_X2, key2)
    return calculated_ciphertext1 == ciphertext_16bit1 and calculated_ciphertext2 == ciphertext_16bit2


## Task 2B

In [133]:
# Example Usage of MITM Attack with TWO P-C pairs

# 1. Generate a random key
true_key_16bit = generate_random_key_16bit()

# 2. Generate TWO plaintext-ciphertext pairs using the same key
plaintext_example1 = 0xABCD
ciphertext_example1 = encrypt(plaintext_example1, true_key_16bit)

plaintext_example2 = 0x5678 # A different plaintext
ciphertext_example2 = encrypt(plaintext_example2, true_key_16bit)


print("True Key:", f"0x{true_key_16bit:04X}")
print("\nPlaintext 1:", f"0x{plaintext_example1:04X}")
print("Ciphertext 1:", f"0x{ciphertext_example1:04X}")
print("\nPlaintext 2:", f"0x{plaintext_example2:04X}")
print("Ciphertext 2:", f"0x{ciphertext_example2:04X}")

# 3. Perform MITM attack using the TWO plaintext-ciphertext pairs
print("\nStarting Meet-in-the-Middle Attack (using two P-C pairs)...")
potential_key_pairs = mitm_attack(plaintext_example1, ciphertext_example1, plaintext_example2, ciphertext_example2)

# 4. Verify potential key pairs against BOTH P-C pairs
print("\nStep D: Verifying potential key pairs against both P-C pairs...")
if potential_key_pairs:
    print(f"Found {len(potential_key_pairs)} potential key pairs. Verifying...")
    correct_key_pair_found = None
    for key_pair in potential_key_pairs:
        if verify_key_pair(key_pair, plaintext_example1, ciphertext_example1, plaintext_example2, ciphertext_example2): # Verify against TWO P-C pairs now
            correct_key_pair_found = key_pair
            break # Assuming we are looking for just one correct key pair, stop after finding the first one
    if correct_key_pair_found:
        key1_found, key2_found = correct_key_pair_found
        print("\n!!!!!! MITM Attack Successful !!!!!!")
        print("Recovered Key Pair:")
        print(f"  Recovered Key1: 0x{key1_found:04X}")
        print(f"  Recovered Key2: 0x{key2_found:04X}")

        # To reconstruct the original key (simplified key expansion is used):
        round_keys_true = key_expansion(true_key_16bit)
        true_key1 = round_keys_true[1]
        true_key2 = round_keys_true[2]
        print("\nVerification with True Keys:")
        print(f"  True Key1:      0x{true_key1:04X}")
        print(f"  True Key2:      0x{true_key2:04X}")

        if key1_found == true_key1 and key2_found == true_key2:
            print("\n!!!!!! Recovered Key Pair matches the true key pair !!!!!!")
        else:
            print("\n?????????? Recovered Key Pair does NOT match the true key pair (but is still a valid key for the given P-C pairs) ??????????")
    else:
        print("\n?????????? MITM Attack Failed: No key pair verified against both P-C pairs (unexpected) ??????????")
else:
    print("\n MITM Attack Failed: No potential key pairs found.")

True Key: 0xB149

Plaintext 1: 0xABCD
Ciphertext 1: 0x9AF8

Plaintext 2: 0x5678
Ciphertext 2: 0x5B1C

Starting Meet-in-the-Middle Attack (using two P-C pairs)...
Step A: Pre-calculating X values (using P1)...
Step B: Pre-calculating X' values (using C1)...
Step C: Finding matching key pairs...
Step C: Found 65536 potential key pairs.

Step D: Verifying potential key pairs against both P-C pairs...
Found 65536 potential key pairs. Verifying...

!!!!!! MITM Attack Successful !!!!!!
Recovered Key Pair:
  Recovered Key1: 0x7872
  Recovered Key2: 0xC632

Verification with True Keys:
  True Key1:      0x7872
  True Key2:      0xC632

!!!!!! Recovered Key Pair matches the true key pair !!!!!!


## Task 3A

What is the key space for the mini block cipher?


*   The key size for the mini block cipher is 16 bits.
*   This means there are 2^16 or 65,536 possible keys.
*   Since we have two keys (Key1 and Key2), the total key space is:
  *   2^16 × 2^16 = 2^32 = 4,294,967,296 key pairs






## Task 3B

If the mini block cipher is executed twice (double mini block cipher), how many operations are needed for a MITM attack?


* Now, we have a 32-bit key, meaning the key space grows to 2^32.
* The MITM attack splits the key search into two parts:
  * One part encrypts all possible 2^16 values for Key1.
  * The second part decrypts all possible 2^16 values for Key2.
* Number of operations:
  * 2^16 + 2^16 = 2^17 = 131,072 operations

## Task 3C

 If we do exhaustive key search for the double mini block cipher, how many operations are needed?

* Exhaustive key search means testing all 2^32 possible keys.
* Number of operations:
  * 2^32 = 4,294,967,296 operations

## Task 3D

Trade-offs between Meet-in-the-Middle Attack and Exhaustive Key Search

The **Meet-in-the-Middle (MITM) attack** and **exhaustive key search** have distinct trade-offs in terms of speed and memory usage.

#### **1. Speed Trade-off**
- **MITM Attack:** Requires **\(2^{17}\)** operations, which is significantly faster than brute-force.
- **Exhaustive Key Search:** Requires **\(2^{32}\)** operations, making it infeasible for larger key sizes.
- **Advantage:** MITM is much **faster** than brute-force, making it more practical.

#### **2. Memory Trade-off**
- **MITM Attack:** Requires storing **\(2^{16}\)** intermediate values in a lookup table.
- **Exhaustive Key Search:** Requires almost no additional memory beyond key testing.
- **Advantage:** Brute-force requires **less memory**, whereas MITM **trades memory for speed**.

#### **3. Applicability Trade-off**
- **MITM Attack:** Only works for ciphers with a **specific structure** (e.g., two independent rounds).
- **Exhaustive Key Search:** Works for **any** cipher.
- **Advantage:** Brute-force is **more general**, but MITM is **more efficient when applicable**.

#### **Conclusion**
The **MITM attack** is significantly **faster** than exhaustive search, but it requires **additional memory** to store intermediate values. It is a **practical attack** when applicable but does not work on ciphers specifically designed to resist it (e.g., AES). On the other hand, **exhaustive search works universally but is computationally infeasible for large key sizes.**